File: Homework/Quang_Minh/HomeWorkLesson9/ValidateBinarySearchTree.java
package HomeWorkLesson9;

import HomeWorkLesson7.TreeNode;

public class ValidateBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
        return isValid(root, null, null);
    }

    public boolean isValid(TreeNode u, TreeNode min, TreeNode max) {
        if (u == null) return true;
        if (max != null && u.val >= max.val) return false;
        if (min != null && u.val <= min.val) return false;
        return isValid(u.left, min, u) && isValid(u.right, u ,max);
    }
}

File: Homework/Van_Le/lesson9/ValidateBinarySearchTree.cs
namespace LeetCodeSolutions
{
    public class ValidateBinarySearchTree
    {
        private int? prev;

        public bool IsValidBST(TreeNode root)
        {
            prev = null;
            return InOrder(root);
        }

        private bool InOrder(TreeNode root)
        {
            if (root == null)
            {
                return true;
            }
            if (!InOrder(root.left))
            {
                return false;
            }
            if (prev != null && root.val <= prev)
            {
                return false;
            }
            prev = root.val;
            return InOrder(root.right);
        }
    }
}
